lib/lfrfid/lfrfid_raw_worker.c

```diff
@@ -0,0 +1,602 @@
+        if(len != sizeof(uint32_t)) {
+            ctx->overrun_count++;
+        }
+    }
+}
+
+static bool read_pair(LFRFIDRawWorkerEmulateData* data, uint32_t* duration, uint32_t* pulse) {
+    RfidFileBuffer* buffer = &data->buffer;
+
+    size_t length = 0;
+    if(buffer->counter >= buffer->size) {
+        if(stream_eof(data->stream)) {
+            // rewind stream and pass header
+            stream_seek(
+                data->stream, strlen(RAW_HEADER) + sizeof(uint32_t), StreamOffsetFromStart);
+        }
+
+        length = stream_read(data->stream, (uint8_t*)&buffer->size, sizeof(size_t));
+        if(length != sizeof(size_t)) {
+            FURI_LOG_E(TAG_EMULATE, "read pair: failed to read size");
+            return false;
+        }
+
+        if(buffer->size > RFID_DATA_BUFFER_SIZE) {
+            FURI_LOG_E(TAG_EMULATE, "read pair: buffer size is too big");
+            return false;
+        }
+
+        length = stream_read(data->stream, buffer->data, buffer->size);
+        if(length != buffer->size) {
+            FURI_LOG_E(TAG_EMULATE, "read pair: failed to read data");
+            return false;
+        }
+
+        buffer->counter = 0;
+    }
+
+    uint32_t value1 = 0;
+    uint32_t value2 = 0;
+
+    buffer->counter += varint_uint32_unpack(
+        &value1, &buffer->data[buffer->counter], (size_t)(buffer->size - buffer->counter));
+
+    if(buffer->counter >= buffer->size) {
+        FURI_LOG_E(TAG_EMULATE, "read pair: buffer is too small");
+        return false;
+    }
+
+    buffer->counter += varint_uint32_unpack(
+        &value2, &buffer->data[buffer->counter], (size_t)(buffer->size - buffer->counter));
+
+    *duration = value2 / 8;
+    *duration -= 1;
+    *pulse = value1 / 8;
+
+    return true;
+}
+
+static int32_t lfrfid_raw_emulate_worker_thread(void* thread_context) {
+    LFRFIDRawWorker* worker = thread_context;
+
+    bool file_valid = true;
+
+    LFRFIDRawWorkerEmulateData* data = malloc(sizeof(LFRFIDRawWorkerEmulateData));
+
+    Storage* storage = furi_record_open("storage");
+    data->stream = file_stream_alloc(storage);
+    data->ctx.overrun_count = 0;
+    data->ctx.stream = xStreamBufferCreate(sizeof(uint32_t), sizeof(uint32_t));
+    data->buffer.size = 0;
+    data->buffer.counter = 0;
+
+    do {
+        const char* filename = string_get_cstr(worker->file_path);
+
+        if(!file_stream_open(data->stream, filename, FSAM_READ, FSOM_OPEN_EXISTING)) {
+            file_valid = false;
+            break;
+        }
+
+        // read header
+        char* header_text = malloc(strlen(RAW_HEADER) + 1);
+
+        size_t size = stream_read(data->stream, (uint8_t*)header_text, strlen(RAW_HEADER));
+        if(size != strlen(RAW_HEADER) || strcmp(header_text, RAW_HEADER) != 0) {
+            file_valid = false;
+        }
+
+        free(header_text);
+
+        uint32_t max_buffer_size;
+        size = stream_read(data->stream, (uint8_t*)&max_buffer_size, sizeof(uint32_t));
```

Comment:
same code in CLI loader, move to 1 place?

Reply:
Moved